docs/madr/decisions/057-rework-mesh-routes-matching.md

```diff
@@ -0,0 +1,922 @@
+
+When referencing Mesh*Route we should validate policy configuration to make sure user is applying only configuration 
+that can be applied on route.
+
+### Affected policies
+
+Policies that can be configured per route: 
+
+- `MeshTimeout` - at the moment only implemented policy, we should enforce validation to only accept config that can be set per route.  
+- `MeshRetry` - would need to be implemented. Like timeout should only accept route related config.
+- `MeshLoadBalancingStrategy` - would need to be implemented. Like timeout should only accept route related config.
+- `MeshRateLimit` - while it's possible to set rate limit on a route. We currently only set rate limits on the inbound so 
+supporting `spec.to[].target.kind: Mesh*Route` doesn't make sense.
+- `MeshAccessLog` - would need to be implemented, with access log filter
+
+#### Producer/consumer model for targeting Mesh*Route in other policies
+
+For MeshRetry our producer/consumer makes sense since producer config is a suggestion from owner of the service. If consumer
+wants to override it because they don't want to retry, they just want to fail fast or they want to retry on bigger number of errors
+they can configure it by applying consumer policy.
+
+MeshRetry example:
+
+```yaml
+# Producer policy with retry on 500 HTTP error code
+apiVersion: kuma.io/v1alpha1
+kind: MeshRetry
+metadata:
+  name: producer-retry
+  namespace: backend-ns
+  labels:
+    kuma.io/mesh: default
+spec:
+  to:
+    - targetRef:
+        kind: MeshHTTPRoute
+        name: backend-route # backend-route must be a producer route, for it to be synced to other zones and work
+      default:
+        http:
+          numRetries: 3
+          backOff:
+            baseInterval: 10ms
+            maxInterval: 1s
+          retryOn:
+            - "500"
+---
+# Consumer policy with override to retry on all 5xx errors
+apiVersion: kuma.io/v1alpha1
+kind: MeshRetry
+metadata:
+  name: consumer-retry
+  namespace: frontend-ns
+  labels:
+    kuma.io/mesh: default
+spec:
+  to:
+    - targetRef:
+        kind: MeshHTTPRoute
+        name: backend-route
+        namespace: backend-ns
+      default:
+        http:
+          retryOn:
+            - "5xx"
+```
+
+For MeshLoadBalancingStrategy adding producer/consumer model can be problematic, for load balancing and hashing. Since this
+is configured on outbound it falls into producer/consumer policy model. But who should have more power on deciding hashing rules?
+It should be producer who is an owner of service who decides how to apply this configuration which is not compliant
+with our consumer/producer model. 
+
+For example, we could have a service that each instance is responsible for handling specific clients. We can use `RingHash`
+to make sure that specific clients requests are processed by the same instance. This should be configured in producer policy. 
+Changing this at consumer level can influence correctness of computations in our service.
```

Comment:
I think it's fine. Consumers can do stupid things but it doesn't matter becasue they only impact themselves.

Reply:
updated